### PR TITLE
[dynamo][guards-cpp-refactor] GetAttrGuardAccessor

### PR DESCRIPTION
**Pull request by anijain2305**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #120123
* #120119
* #120098
* #120096
* #120093
* #120091
* #120089
* #120068
* #120067
* #120065
* #120064
* #120062
* #120061
* #120060
* __->__ #119833
* #119827
* #119822

cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @chenyang78 @aakhundov @kadeng

### CODE REVIEW

**Review by jansel — APPROVED** [review]

File: torch/csrc/dynamo/guards.cpp [hunk 1]
```diff
@@ -1318,6 +1318,60 @@ std::unique_ptr<GuardManager> make_guard_manager(
...
+  // check_verbose_nopybind.
+  bool check_nopybind(PyObject* obj) override { // borrowed ref
+    PyObject* x = PyObject_GetAttr(obj, _attr_name); // new ref
+    if (x == nullptr) {
+      // Attribute absent, clear the exception and return false.
+      PyErr_Clear();
+      return false;
+    }
+    bool result = _guard_manager->check_nopybind(x);
+    Py_DECREF(x);
```
[the remark this answers is not in the document]
Comment (anijain2305) [reply torch/csrc/dynamo/guards.cpp:1357]: I don't fully understand the question.

1) If an exception is raised in L1350, we clear it. As x is `nullptr`, we don't need to DECREF it.
2) If an exception is raised in L1356 in a child guard manager, there are two cases
      a) Do we clear the exception like `PyErr_Clear()`?
      b) Or do we not clear it? (like in LAMBDA_GUARD, we don't clear the exception but we return false)
In both 2a and 2b, we will still reach L1357 and DECREF. 

So, I think we are safe here?

**Pull request closed without merging**